### PR TITLE
Refactor yarpcgrpc to match new errors API

### PR DESCRIPTION
**Pull request by mh-park**

### CODE REVIEW

File: v2/yarpcgrpc/handler.go [hunk 1]
```diff
@@ -246,3 +246,32 @@ func handlerErrorToGRPCError(err error, mdWriter *metadataWriter) error {
...
+	if info != nil {
+		if info.Name != "" {
+			mdWriter.AddSystemHeader(ErrorNameHeader, info.Name)
+		}
+		if resBuf != nil {
+			switch encoding {
+			case yarpcprotobuf.Encoding:
+				mdWriter.AddSystemHeader(ErrorDetailsHeader, resBuf.String())
+			case yarpcjson.Encoding, yarpcthrift.Encoding:
+				mdWriter.AddSystemHeader(ApplicationErrorHeader, resBuf.String())
```
Comment (robbertvanginkel) [inline v2/yarpcgrpc/handler.go]: Can we now remove `ApplicationErrorHeaderValue` from `v2/yarpcgrpc/headers.go:81`? According to the comment above that, it says that `rpc-application-header`'s value represents an error if the value is non-empty. What definition does this refer to, yarpc is this a yarpc definition?

File: v2/yarpcgrpc/handler.go [hunk 2]
```diff
@@ -246,3 +246,32 @@ func handlerErrorToGRPCError(err error, mdWriter *metadataWriter) error {
 	}
 	return status.Error(grpcCode, message)
 }
+
+func handleResponse(
+	encoding yarpc.Encoding,
+	info *yarpcerror.Info,
```
Comment (robbertvanginkel) [inline v2/yarpcgrpc/handler.go]: maybe `errInfo` to make make its usage more apparent.

File: v2/yarpcgrpc/handler.go (hunk 1 above)
Comment (mh-park) [reply v2/yarpcgrpc/handler.go]: Yea, it looks like this header value wasn't actually used anywhere except as a placeholder. On the outbound side, the client would only check if `ApplicationErrorHeader` exists. In v2, I'd like to use this header to actually hold the serialized error in JSON/Thrift's case.

I might be missing something, @AlexAPB, @kriskowal is this correct and is it safe for me to take this approach?

File: v2/yarpcgrpc/handler.go [hunk 3]
```diff
@@ -246,3 +246,32 @@ func handlerErrorToGRPCError(err error, mdWriter *metadataWriter) error {
...
+	if errInfo == nil && resBuf != nil {
+		return serverStream.SendMsg(resBuf.Bytes())
+	}
+
+	// This is an application error
+	if errInfo != nil {
+		if errInfo.Name != "" {
+			mdWriter.AddSystemHeader(ErrorNameHeader, errInfo.Name)
+		}
+		if resBuf != nil {
```
Comment (AlexAPB) [inline v2/yarpcgrpc/handler.go]: this should never be `nil`, correct?

File: v2/yarpcgrpc/handler.go [hunk 4]
```diff
@@ -246,3 +246,32 @@ func handlerErrorToGRPCError(err error, mdWriter *metadataWriter) error {
...
+			switch encoding {
+			case yarpcprotobuf.Encoding:
+				mdWriter.AddSystemHeader(ErrorDetailsHeader, resBuf.String())
+			case yarpcjson.Encoding, yarpcthrift.Encoding:
```
Comment (AlexAPB) [inline v2/yarpcgrpc/handler.go]: Can we replace this with a `default` case?

File: v2/yarpcgrpc/handler.go (hunk 1 above)
Comment (AlexAPB) [reply v2/yarpcgrpc/handler.go]: for backwards compatibility, we'll need this header to be set whenever there's an application error. This also goes for the above protobuf statement.

File: v2/yarpcgrpc/headers.go [hunk 5]
```diff
@@ -66,17 +66,13 @@ const (
 	EncodingHeader = "rpc-encoding"
 	// ErrorNameHeader is the header key for the error name.
 	ErrorNameHeader = "rpc-error-name"
+	// ErrorDetailsHeader is the header key for the error details used in gRPC.
+	// https://sourcegraph.com/github.com/grpc/grpc-go/-/blob/internal/transport/http_util.go#L317-326
```
Comment (AlexAPB) [inline v2/yarpcgrpc/headers.go]: if this is the best we've got, let's use a permalink instead (includes the hash)

File: v2/yarpcgrpc/headers.go [hunk 6]
```diff
@@ -66,17 +66,13 @@ const (
...
+	ErrorDetailsHeader = "grpc-status-details-bin"
 	// ApplicationErrorHeader is the header key that will contain a non-empty value
 	// if there was an application error.
 	ApplicationErrorHeader = "rpc-application-error"
```
Comment (kriskowal) [inline v2/yarpcgrpc/headers.go]: We can remove ApplicationErrorHeader and ApplicationErrorHeaderValue safely, or at least leave their behavior unchanged and orthogonal to the grpc status details binary header, for compatibility with existing YARPC gRPC services. We need to engage with the gRPC community separately to find a different header than grpc-status-details-bin for non-Protobuf error details. We don’t have to validate gRPC+Thrift yet so we can leave this unimplemented for now.

File: v2/yarpcgrpc/handler.go (hunk 3 above)
Comment (mh-park) [reply v2/yarpcgrpc/handler.go]: not necessarily; take json for example: if they specify a yarpcerror with no details in their user handler, we don't put anything in the buffer.

we could make it always have a non-nil value by putting an empty buffer there though.

**Pull request merged**